### PR TITLE
Bookmarking feature

### PR DESCRIPTION
**Pull request by justinctlam**
- Added support for fetching bookmarks in a new bookmarks tab
- Added support for bookmarking a pod cast
- Added test hook debugging tools in a new debugging tab

### CODE REVIEW

**Comment by justinctlam** [comment]
There is still one issue/polish to be fixed but it might involve some refactoring of the generic collection view. The biggest issues is that the generic collection view doesn't refresh when changes are made in any other collection views including the bookmark collection view. e.g. up vote, down vote or bookmarking in one collection view will not be reflected in another until app restart.

Also minor polishes include bookmarking doesn't have paging support from the API side and paging for posts should be cleaned up.

**Comment by themisterholliday** [comment]
I've been working on a fix for multiple datasources being updated at once. So I'll see if I can't add that in before we merge this in so everything can work properly.

**Comment by justinctlam** [comment]
Just fyi, I hope this doesn't sit around too long. I'm afraid of merging changes after not looking at this for over a month.

**Comment by themisterholliday** [comment]
Yea sorry. I've fallen behind on checking PR's. I'll merge this in today or tomorrow. This shouldn't break anything and if it does I'll fix it real quick.

**Comment by TheHollidayInn** [comment]
@justinctlam would you mind fixing the merge conflicts on this PR?

**Comment by justinctlam** [comment]
I have fixed the merge conflicts. Please give it a try.

**Comment by TheHollidayInn** [comment]
Thank you!

**Comment by TheHollidayInn** [comment]
Alright! Looks pretty good. Couple of things to note:
- Can we move the bookmark code into its own folder `Bookmark` just like the debug and search code.
- After bookmarking my first item, then going to the bookmark tab, I have to pull to refresh before seeing my book marks. Can you make view update automatically? I understand there may need to be some refactoring, so we can do that in a separate PR if need be.
- Can we move the bookmark button the the action bar at top right rather then next to the upvote/downvote?
- The debug tab simply crashes the app when I click it

File: SEDaily-IOS/StateBookmarkView.swift [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        self.stackView = UIStackView()
+        self.stackView.axis = .vertical
+        self.addSubview(self.stackView)
+
+        self.label = UILabel(text: text)
+        self.stackView.addArrangedSubview(self.label)
+
+        if showRefreshButton {
+            self.refreshButton = UIButton()
+            self.refreshButton.setTitle("Tap to refresh", for: .normal)
```
Comment (jasonify) [inline SEDaily-IOS/StateBookmarkView.swift]: Maybe add to Localizable.strings

**Comment by justinctlam** [comment]
> Can we move the bookmark code into its own folder Bookmark just like the debug and search code.

Fixed!

> After bookmarking my first item, then going to the bookmark tab, I have to pull to refresh before seeing my book marks. Can you make view update automatically? I understand there may need to be some refactoring, so we can do that in a separate PR if need be.

I understand. There's also another issue where you bookmark something in the bookmark view and the main podcast view doesn't update. Is Keith doing some changes in the repository and caching architecture?

> Can we move the bookmark button the the action bar at top right rather then next to the upvote/downvote?

Can you give me a design layout for this? I don't quite understand how it should look

> The debug tab simply crashes the app when I click it

Apologies for this, seems like a bad merge from before. It should be fixed now.

**Comment by TheHollidayInn** [comment]
Thanks for fixing so fast!

> I understand. There's also another issue where you bookmark something in the bookmark view and the main podcast view doesn't update. Is Keith doing some changes in the repository and caching architecture?

I'm Keith, fyi :P But Craig did show me this issue and how to fix it. So I can double check then create a PR, so let's leave this for a separate PR. We will just do it before pushing live.

> Can you give me a design layout for this? I don't quite understand how it should look

I used this example in the share PR as well. If you look at the image on the right, there are action buttons at the top.

https://i.amz.mshcdn.com/Ukg_i3vyZB2DFuJToFqcjqUt0WM=/fit-in/1200x9600/http%3A%2F%2Fmashable.com%2Fwp-content%2Fuploads%2F2015%2F04%2FApple-Watch-App-Store-1-640x565.jpg

**Comment by justinctlam** [comment]
Yes thanks Keith, I'm getting the two mixed up ;)

I've updated the bookmark location and fixed a localization text.

**Comment by TheHollidayInn** [comment]
Works great! I'm checking with Craig before merging so I don't interfere with a release.

**Pull request merged**